sycl/doc/design/MappingHostAddressesToDeviceEntities.md

```diff
@@ -0,0 +1,369 @@
+host compiler.
+
+Both approaches have their pros and cons and they are expected to be implemented
+and exist in the implementation at the same time. Only one of them will be
+used at a time, depending on whether a 3rd-party host compiler is used or not.
+
+Integration footer can be used with 3rd-party host compilers. This, however
+requires appending to a translation unit provided by a user, which could affect
+debug information. Since there are no compilers that support appending a file at
+the end (similar to `-include`), appending is done by generating a temporary
+input file using concatenation of the original input and integration footer.
+
+Such replacement of the main translation unit causes the following issues:
+- debug information about the source file might be incorrect, leading to
+  problems with gdb `l` command and code coverage tools
+- checksum of host and device source files becomes different which causes device
+  code debugging to be completely broken in some environments (such as MS Visual
+  Studio, for example)
+
+Customizing host compiler allows to avoid issues with debuggers and code
+coverage tools, but that is not an option if a user wants to compile host part
+of an app with a 3rd-party host compiler.
+
+The sections below describe the implementation design of both approaches in more
+detail. Note that there are few components which should be modified regardless
+of which approach is in use.
+
+## Common front-end part
+
+DPC++ FE should support the following attribute:
+`[[__sycl_detail__::uniquely_identifiable_object(kind)]]`. The attribute accepts
+a string literal and should be applied to types (like `device_global` or
+`specialization_id`).
+
+Presence of the attribute instructs the compiler to perform the following
+things:
+- emit `sycl-unique-id` LLVM IR attribute on each definition of a variable of
+  type marked with `[[__sycl_detail__::uniquely_identifiable_object(kind)]]`
+  attribute. `sycl-unique-id` LLVM IR attribute should be accompanied by a
+  unique string identifier of the variable it is attached to. The rules for
+  creating this string are the same as for `__builtin_sycl_unique_stable_id` and
+  the same algorithm can be used when generating the string for the attribute
+- emit `sycl-uid-kind` LLVM IR attribute alongside `sycl-unique-id`, which
+  contains the `kind` string passed via
+  `[[__sycl_detail__::uniquely_identifiable_object(kind)]]` attribute
+
+To illustrate, here is a SYCL code snippet:
+
+```
+template <typename T>
+class
+  // Note: the attribute usage will be guarded by macro to be only applied when
+  // DPC++ compiler is used to avoid generating warnings. That is described
+  // later in the doc
+  [[__sycl_detail__::uniquely_identifiable_object("specialization_id")]]
+  specialization_id {
+  // ...
+};
+
+specialization_id<int> spec_const(38);
+```
+
+After processed by DPC++ compiler, it will result in the following LLVM IR:
```

Comment:
You mean LLVM IR for device?

Reply:
I mean for both host and device: there is a paragraph a bit below, which says that. Please let me know how to better rearrange and/or rephrase that to make it more clear

Reply:
Maybe we should mention at the very beginning of attributes description that they trigger emission of new data for both host and device code. This a unique situation among SYCL attributes, so I kind of assumed that it will happen only for device code.